### ISSUE TITLE
Aborted (core dumped) when banning

### ISSUE THREAD

**Issue by shtse8** [coinfoundry/miningcore#9]
```shell
dotnet: /datadrive/teamcity/6a56927d0d6f3630/submodules/libuv/src/unix/stream.c:933: void uv__write_callbacks(uv_stream_t *): Assertion `(((const QUEUE *) (&(stream->loop)->active_reqs) == (const QUEUE *) (*(QUEUE **) &((*(&(stream->loop)->active_reqs))[0]))) == 0)' failed.
Aborted (core dumped)
```

**Comment by oliverw** [comment]
Could you post the full exception information from the log if possible? I need the callstack. Thanks!

**Comment by shtse8** [comment]
I am using Ubuntu 16.04
I can't find the callstack even I enabled the logging to debug and the log file per pool.
This is the last message I have seen before the crash.
```
7-09-24 06:47:45.2721] [D] [dash] [Pool] [0HL832O0Q8N97] Received request mining.submit [550]
[2017-09-24 06:47:45.2721] [D] [JsonRpcConnection] [0HL832O0Q8N96] Sending response: {"result":{"code":23,"message":"low difficulty share (18.868918098953)","data":null},"error":null,"id":1693}
[2017-09-24 06:47:45.2721] [D] [dash] [Pool] [0HL832O0Q8N96] Received request mining.submit [1694]
[2017-09-24 06:47:45.2721] [D] [dash] [Pool] [0HL832O0Q8N97] Updating VarDiff
[2017-09-24 06:47:45.2721] [D] [dash] [Pool] [0HL832O0Q8N96] Updating VarDiff
[2017-09-24 06:47:45.2721] [D] [JsonRpcConnection] [0HL832O0Q8N97] Sending response: {"result":{"code":23,"message":"low difficulty share (41.6248068372879)","data":null},"error":null,"id":550}
[2017-09-24 06:47:45.2721] [D] [JsonRpcConnection] [0HL832O0Q8N96] Sending response: {"result":{"code":23,"message":"low difficulty share (13.4755671123327)","data":null},"error":null,"id":1694}
[2017-09-24 06:47:45.2721] [D] [dash] [Pool] [0HL832O0Q8N96] Received request mining.submit [1695]
[2017-09-24 06:47:45.2721] [D] [dash] [Pool] [0HL832O0Q8N96] Updating VarDiff
[2017-09-24 06:47:45.2774] [D] [JsonRpcConnection] [0HL832O0Q8N96] Sending response: {"result":{"code":23,"message":"low difficulty share (19.0935004327123)","data":null},"error":null,"id":1695}
[2017-09-24 06:47:45.2774] [D] [dash] [Pool] [0HL832O0Q8N96] Received request mining.submit [1696]
[2017-09-24 06:47:45.2774] [D] [dash] [Pool] [0HL832O0Q8N96] Updating VarDiff
[2017-09-24 06:47:45.2774] [D] [JsonRpcConnection] [0HL832O0Q8N96] Received JsonRpc-Request: {"id":1888,"method":"mining.submit","params":["XrQsEqFfabd5mGy7W9vjBnvk6K19oc97gB.nicehash","22","9c540100","59c754da","b1ce3dd6"]}
[2017-09-24 06:47:45.2774] [D] [JsonRpcConnection] [0HL832O0Q8N96] Received JsonRpc-Request: {"id":1889,"method":"mining.submit","params":["XrQsEqFfabd5mGy7W9vjBnvk6K19oc97gB.nicehash","22","ce3d0100","59c754da","d27daea1"]}
[2017-09-24 06:47:45.2774] [W] [dash] [Pool] [0HL832O0Q8N97] Banning worker for 600 sec: 100% of the last 51 shares were invalid
[2017-09-24 06:47:45.2774] [D] [JsonRpcConnection] [0HL832O0Q8N96] Sending response: {"result":{"code":23,"message":"low difficulty share (13.6891748137657)","data":null},"error":null,"id":1696}
[2017-09-24 06:47:45.2774] [W] [dash] [Pool] [0HL832O0Q8N96] Banning worker for 600 sec: 100% of the last 51 shares were invalid
[2017-09-24 06:47:45.2802] [D] [dash] [Pool] [0HL832O0Q8N97] Received request mining.submit [551]
[2017-09-24 06:47:45.2802] [D] [JsonRpcConnection] [0HL832O0Q8N97] Last subscriber disconnected from receiver stream
dotnet: /datadrive/teamcity/6a56927d0d6f3630/submodules/libuv/src/unix/stream.c:933: void uv__write_callbacks(uv_stream_t *): Assertion `(((const QUEUE *) (&(stream->loop)->active_reqs) == (const QUEUE *) (*(QUEUE **) &((*(&(stream->loop)->active_reqs))[0]))) == 0)' failed.
Aborted (core dumped)
```

**Comment by oliverw** [comment]
Anything in <code>core.log</code>?

**Comment by shtse8** [comment]
nothing special.
```
[2017-09-24 06:47:45.2774] [W] [dash] [Pool] [0HL832O0Q8N97] Banning worker for 600 sec: 100% of the last 51 shares were invalid 
[2017-09-24 06:47:45.2774] [D] [JsonRpcConnection] [0HL832O0Q8N96] Sending response: {"result":{"code":23,"message":"low difficulty share (13.6891748137657)","data":null},"error":null,"id":1696} 
[2017-09-24 06:47:45.2774] [W] [dash] [Pool] [0HL832O0Q8N96] Banning worker for 600 sec: 100% of the last 51 shares were invalid 
[2017-09-24 06:47:45.2802] [D] [dash] [Pool] [0HL832O0Q8N97] Received request mining.submit [551] 
[2017-09-24 06:47:45.2802] [D] [JsonRpcConnection] [0HL832O0Q8N97] Last subscriber disconnected from receiver stream 
```

For my best guess, it should be crashed when miningcore ban any client. and the banning mechinisam is not good for pool farm like nicehash as it is now checking the last 51 shares. as nicehash got lots of hashpower and it sometimes submitting shares too quick no matter vardiff or staticdiff. so it will always want to ban nicehash. we can't buy stable hashpower from nicehash because we need to bid against others. static diff is hard to adjust for nicehash, vardiff need sometimes to adjust.

**Comment by shtse8** [comment]
I disbled the banning and no crashs now.

**Comment by oliverw** [comment]
Interesting observations. I will investigate further. Thanks a lot for the feedback.

**Comment by shtse8** [comment]
thanks.
it now have two problems.
1. the banning mechinisam for pool farm and unstble hashrate
2. crash when banning client in Ubuntu 16.04

**Comment by oliverw** [comment]
I've just checked in a potential fix. Could you verify the problem persists using this change?

**Comment by shtse8** [comment]
Great. Let me check but give me some time.
because i need to create an order and wait for low price in Nicehash. 
I will reply later.

**Comment by shtse8** [comment]
I have tested with the latest commit. Crash with `Aborted (core dumped)` once, then I want to turn on the logging. Until now, nothing happen. But it seems there is a serious problem when a client submitting shares too fast. Please check https://github.com/coinfoundry/miningcore/issues/11 . I know turning diff can avoid the problem. But for pool farm with the cheapest orders. sometimes we only get too little hashpower. but sometimes it gives us lots of hashpower. it's hard to adjust.

**Comment by oliverw** [comment]
Could this have been resolved by the VarDiff changes?

**Comment by shtse8** [comment]
No `Aborted` occur now, but `Segmentation Fault` when running for several hours.

**Comment by shtse8** [comment]
Aborted Again
```
[2017-09-27 07:00:33.2121] [I] [dgb-scrypt] [Pool] [0HL85ABLCB07Q] Booting zombie-worker (idle-timeout exceeded)
[2017-09-27 07:00:33.2121] [I] [dgb-scrypt] [Pool] [0HL85ABLCB07O] Booting zombie-worker (idle-timeout exceeded)
dotnet: /datadrive/teamcity/6a56927d0d6f3630/submodules/libuv/src/unix/stream.c:933: void uv__write_callbacks(uv_stream_t *): Assertion `(((const QUEUE *) (&(stream->loop)->active_reqs) == (const QUEUE *) (*(QUEUE **) &((*(&(stream->loop)->active_reqs))[0]))) == 0)' failed.
Aborted (core dumped)
```

**Comment by oliverw** [comment]
Fixed by 6e35a735f611043ae8862bfbeff566e75753a33c
